browser/brave_wallet/BUILD.gn

```diff
@@ -3,21 +3,24 @@ import("//extensions/buildflags/buildflags.gni")
 
 assert(brave_wallet_enabled)
 
-source_set("brave_wallet") {
+source_set("brave_wallet_impl") {
```

Comment:
We'd better document somehow why we are keeping both `source.gni` and `BUILD.gn` to prevent chaos and confusion.

Basically maybe @mariospr you could write a one-pager describing our gn fixing effort and its final goal? Because people already get confused with how the code should be structured

Reply:
> We'd better document somehow why we are keeping both source.gni and BUILD.gn to prevent chaos and confusion.
>
> Basically maybe @mariospr you could write a one-pager describing our gn fixing effort and its final goal? Because people already get confused with how the code should be structured

I'm happy to do this, but I'd wait a bit if it's ok until this PR lands and then the one I'm preparing since that way I think it would be easier to explain without creating even more confusion, since the exact way these fixes are implementing are yet kind of a moving target at the moment.

Reply:
part of the problem is that things are so entangled right, but ideally we only want to use sources.gni for things that would otherwise have circular dependencies. Currently we have almost everything in sources.gni as a stepping stone

Reply:
what is the difference in managing deps?

I'm not into idea of maintaining two approaches, because: 
1. not clear when to use one or another 
2. eventually one will have to refactor `build.gn` to `sources.gni` when something unrelated changes outside of that part (e.g. in chromium) 
3. once we have `sources.gni` for something nobody will transform it back to `build.gn` even if the circular dep doesn't exist anymore

Reply:
I think it's very clear when to use one or the other. You only use sources.gni when there is a circular dep with upstream code, in all other cases you should use normal targets. It's the same logic you apply when deciding whether to create a new target or include files in existing targets. In most cases you should try to have one BUILD.gn for each directory. If you're adding a file that needs to be part of an existing brave target you just add it. If you're adding a file that needs to be a part of an existing chrome target, you use sources.gni. 

Also it's not true that nobody will change it back to BUILD.gn, that is something @mariospr is actively working on, but it was essentially impossible to untangle things without this first step. We have tickets and have been actively working on removing `check_includes = false` and this is part of that work. We certainly don't want huge lists of files in one target because that will negatively impact incremental builds among other things. It also doesn't work for components because there's no existing upsteam target to add the sources to. Upstream is slowly fixing targets to more closely match ios/chrome which doesn't have the giant targets that desktop/android inherited from gyp.

Reply:
I'm breaking this PR into smaller pieces so this is being removed for now, but I think it's really important that we don't use sources.gni for everything. It should only be used when we need to insert sources into existing upstream targets

Reply:
I'm pretty much aware of tickets and `check_includes = false`, because at some point in the past I pushed for creating all those tickets and for the whole initiative

> It also doesn't work for components because there's no existing upsteam target to add the sources to.

I never suggested to make `sources.gni` for components

> I think it's very clear when to use one or the other. You only use sources.gni when there is a circular dep with upstream code

It is not, otherwise people would not create a ton of code with circular dependencies and layering violations

> We certainly don't want huge lists of files in one target because that will negatively impact incremental builds among other things

I don't see how this affect incremental builds. Chrome has almost everything in `chrome/browser/BUILD.gn` and it works well. I have some previous experience, and in that browser we also injected all `chrome/browser/xyz` sources (and there were 50x more custom code than in Brave) without any issues with incremental builds.

So I'd still suggest to stick to `source.gni` approach (by default, with possible exceptions) for sources in `brave/browser` for the reasons I've mentioned above: it is more simple, it doesn't have significant downsides and is  more stable vs possible chromium changes (and our own refactorings).

Reply:
People are creating a ton of code with circular dependencies and layering violations because we didn't have gn check running until recently and now we have a bunch of targets with check_includes = false. It's very simple: if the source needs to go in an existing chromium target then it goes in sources.gni, otherwise it goes in its own target.  This is basically like saying that every file should just go in one giant target because no one knows where to put them. I think it does have significant downsides (more patching) and that necessarily makes it less stable with chromium changes. If we don't do more patching then we are dumping everything into a small number of top-level targets and that's exactly the opposite direction that chromium is moving in. I really couldn't disagree more strongly that using `sources.gni` is the way to go for everything in brave/browser. The sole purpose of sources.gni is to eliminate the circular dependencies with a very small number of files.

Reply:
also it affects incremental builds because any change to brave/browser files will trigger a rebuild of the `//chrome/browser` target which in turn will trigger a rebuild of everything that depends on `//chrome/browser` and so on. Smaller targets means fewer files will depend on those targets and therefore fewer files have to be rebuilt when changes are made.

Reply:
Also I don't see this at "two approaches". `sources.gni` is just a replacement for `deps` patches to eliminate the circular dependency issues. Now, I agree it's confusing if some files in the same directory are in BUILD.gn and some are in sources.gni, but that isn't the long-term intention here, it's just a temporary workaround. Maybe it means moving all files that go in sources.gni into chromium_src and I'd be totally fine with that.

Reply:
and actually considering that it's simply not possible to just put all the files in sources.gni in the same upstream target (for instance sources in `//chrome/browser:browser` can't depend on `//chrome/browser/ui`, I think using sources.gni is going to end up with exactly the same kinds of issues with circular deps that are difficult to resolve and can only be fixed with new targets and/or more patching

Reply:
moving this discussion to DM

Reply:
> and actually considering that it's simply not possible to just put all the files in sources.gni in the same upstream target (for instance sources in `//chrome/browser:browser` can't depend on `//chrome/browser/ui`, I think using sources.gni is going to end up with exactly the same kinds of issues with circular deps that are difficult to resolve and can only be fixed with new targets and/or more patching

I'm a bit confused now because after some conversations yesterday I thought that, while we all agree that putting things into sources.gni files for //brave/browser was a necessary intermediate step towards having more reasonable BUILD.gn targets in the end, and I was going to write a small wiki page explaining the current approach we're following, the WIP status of it, and what the end goal would be (as Brian said: things that would cause circular deps going to sources.gni to integrate them with Chrome targets, and everything else into smaller & independent targets).

But this comment (plus another one in Slack) makes me wonder whether we're actually "aborting" that intermediate approach, so I'd appreciate some clarification, thanks :innocent: